### ISSUE TITLE
Linear Extraction Error

### ISSUE THREAD

**Issue by jb10016** [Riverscapes/gcd#318]
An attempt to document the problems I'm having with the LE.  The files are big, so I'll send this first, and see if you can spot an issue I'm blindly ignoring and then move on to projects/databases and videos.

So, I have a large DEM (some 40 km x on average 1 km of river corridor at 2 m resolution) and a network of 200 m sections I am trying to extract.  The section shapefile is attached as a zip file.

The shapefile has a floating point field for the distance and a text label field, as shown below:

![image](https://user-images.githubusercontent.com/35811609/45661721-d9c3cb80-bb52-11e8-9051-2b06060c0d7b.png)

This loads as a longitudinal route without a problem and the GCDFID appears to have been constructed properly.

However, when I attempt a linear extraction, a exception is thrown (details below).  

I'm also having related problems with another dataset, though in that case the sections are extracted ok, but then the database appears to have formatting problems, so I'm not sure if these are related and will leave that for another time.

```
EXCEPTION
Exception has been thrown by the target of an invocation.
 --- Stacktrace --- 
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)
   at System.Reflection.RuntimeConstructorInfo.Invoke(BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.RuntimeType.CreateInstanceImpl(BindingFlags bindingAttr, Binder binder, Object[] args, CultureInfo culture, Object[] activationAttributes, StackCrawlMark& stackMark)
   at System.Activator.CreateInstance(Type type, BindingFlags bindingAttr, Binder binder, Object[] args, CultureInfo culture, Object[] activationAttributes)
   at System.Activator.CreateInstance(Type type, Object[] args)
   at GCDConsoleLib.RasterOperators.MakeGenericType(Type generic, Type innerType, Object[] args) in D:\Code\gcd\gcd\GCDConsoleLib\RasterOperators\Interfaces.cs:line 1246
   at GCDConsoleLib.RasterOperators.GenericRunWithSpacing(Type generic, Type innerType, String sDesc, EventHandler`1 progressHandler, Object[] args) in D:\Code\gcd\gcd\GCDConsoleLib\RasterOperators\Interfaces.cs:line 1308
   at GCDConsoleLib.RasterOperators.LinearExtractor(Vector vLineShp, List`1 rRasters, FileInfo sOutCSV, Decimal intervallength, String sFieldName, EventHandler`1 progressHandler) in D:\Code\gcd\gcd\GCDConsoleLib\RasterOperators\Interfaces.cs:line 965
   at GCDCore.UserInterface.LinearExtraction.frmLinearExtractionProperties.cmdOK_Click(Object sender, EventArgs e)
 --- Exception Data --- 
ArcMap = 10.5.1.7333
GCD = 7.2.0.0Incompatible geometry for operation
 --- Stacktrace --- 
   at OSGeo.OGR.Geometry.GetPoint_2D(Int32 iPoint, Double[] argout)
   at GCDConsoleLib.Internal.Operators.LinearExtractor`1.GetFID0XY() in D:\Code\gcd\gcd\GCDConsoleLib\RasterOperators\Operators\LinearExtractor.cs:line 171
   at GCDConsoleLib.Internal.Operators.LinearExtractor`1..ctor(Vector vLineShp, List`1 rRasters, FileInfo outCSV, Decimal ptsspacing, String FieldName) in D:\Code\gcd\gcd\GCDConsoleLib\RasterOperators\Operators\LinearExtractor.cs:line 151
Windows: Microsoft Windows NT 6.2.9200.0
Date: 18/09/2018 2:46:55 PM
```

[Dart_200m_Sections.zip](https://github.com/Riverscapes/gcd/files/2390970/Dart_200m_Sections.zip)

**Comment by jb10016** [comment]
@philipbaileynar 

I should have added - the shapefile of sections does not coincide entirely with the DEM (i.e., there are sections that lie outside and beyond the DEM) ... but I'm sure this hasn't been a problem in the past.

**Comment by jb10016** [comment]
@philipbaileynar 

For the sake of completeness - the slimmed down, zipped up project is [here](https://drive.google.com/a/waikato.ac.nz/file/d/142LwGr1hvVGLj5OGqYWcX70ftSexJEp3/view?usp=sharing).

**Comment by philipbaileynar** [comment]
@jb10016 the problem appears to revolve around the fact that one of the cross section features in the ShapeFile is a **multipart** feature. You can see this in the following screenshot. Notice how there's one row selected in the attribute table (feature ID 166) but seemingly two geometries selected in the map. In reality these two lines are part of the same feature. The GCD code is choking because it's trying to interogate a single geometry object when it instead encounters an array of geometry objects.

![2018-09-18_142952](https://user-images.githubusercontent.com/8645379/45717679-8d2ac100-bb4f-11e8-8ccc-ac356bcaef9c.png)

We will continue to research a way for us to either handle this situation or better warn the user when data like these are encountered. In the meantime you can avoid the problem by using ArcGIS to split the two lines into two separate features.

**Comment by jb10016** [comment]
@philipbaileynar - Great spot Philip.  Good to know this was a data-driven issue.  It certainly sounds as though a warning message would be the way to go if that is possible?  Thanks for this.

**Comment by philipbaileynar** [comment]
@jb10016 Here's the new warning message. 

![2018-09-18_151023](https://user-images.githubusercontent.com/8645379/45719368-01b42e80-bb55-11e8-8687-d4b92019e2f9.png)

I also ran a quick test. Multipart **polygon** ShapeFiles blow up mask generation (and presumably other features such as budget seg). 

The new warning message above appears everywhere that the user can introduce a Shapefile to a GCD project, so hopefully we have just prevented that polygon case from occurring too.

This warning message is queued and ready for the next GCD release.
